extensions/narayana-jta/deployment/src/main/java/io/quarkus/narayana/jta/deployment/NarayanaJtaProcessor.java

```diff
@@ -65,6 +66,7 @@ public void build(NarayanaJtaRecorder recorder,
         additionalBeans.produce(new AdditionalBeanBuildItem(CDIDelegatingTransactionManager.class));
         runtimeInit.produce(new RuntimeInitializedClassBuildItem(
                 "com.arjuna.ats.internal.jta.resources.arjunacore.CommitMarkableResourceRecord"));
+        runtimeInit.produce(new RuntimeInitializedClassBuildItem(SocketProcessId.class.getName()));
```

Comment:
why dose this class need to be initialized at runtime ?

Reply:
I first tried a fix using BuildProducer<ReflectiveClassBuildItem>:

``` java
reflectiveClass.produce(new ReflectiveClassBuildItem(false, false, SocketProcessId.class.getName(), ...
```

and although the test passed locally it failed on the quarkus CI with a message saying that "... this class should be initialized at image runtime ...":

```
Error: No instances of java.net.SocksSocketImpl are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use -H:+TraceClassInitialization.
Detailed message:
Trace: Object was reached by etc ...
```

So because of the error message I guessed at the fix:

```java
runtimeInit.produce(new RuntimeInitializedClassBuildItem(SocketProcessId.class.getName()));
```

but that fails as well with a different error:

```
Error: Classes that should be initialized at run time got initialized during image building:
 com.arjuna.ats.internal.arjuna.utils.SocketProcessId the class was requested to be initialized at run time (from feature io.quarkus.runner.AutoFeature.beforeAnalysis). To see why com.arjuna.ats.internal.arjuna.utils.SocketProcessId got initialized use -H:+TraceClassInitialization
```

Does anyone have any hints about how to tell the native image build how to process the SocketProcessId class?

Reply:
hmm, the SocketProcessId is looping to create the ServerSocket to find the first available port which should be unique. It makes senses to be initialized at runtime. The second error looks weird and how to add ```-H:+TraceClassInitialization``` option to see what happens ?

Reply:
Hello @zhfeng, 

> how to add -H:+TraceClassInitialization option to see what happens ?

If you want to build the native image locally and test it out with this, then you can do so by passing `-Dquarkus.native.additional-build-args=-H:+TraceClassInitialization` when you run the `mvn` command

Reply:
Thanks @jaikiran - it works !
@mmusgrov it looks like that we have to add ```CommitMarkableResourceRecordRecoveryModule``` and ```RecoverConnectableAtomicAction```
```
runtimeInit.produce(new RuntimeInitializedClassBuildItem(CommitMarkableResourceRecordRecoveryModule.class.getName()));
runtimeInit.produce(new RuntimeInitializedClassBuildItem(RecoverConnectableAtomicAction.class.getName()));
```
But I just wonder how we register CommitMarkableResourceRecordRecoveryModule ? Also the recovery service gets running ?

Reply:
> Hello @zhfeng,
> 
> > how to add -H:+TraceClassInitialization option to see what happens ?
> 
> If you want to build the native image locally and test it out with this, then you can do so by passing `-Dquarkus.native.additional-build-args=-H:+TraceClassInitialization` when you run the `mvn` command

@jaikiran the problem I am having is that the integration test passes locally but fails when it is tested by the PR checks (my previous comment indicates the errors I get when the PR does the checking).

Reply:
Hello @mmusgrov, if this PR has all the necessary changes, then I'll try it out locally tonight. If there are any pending changes you wish to push, please do so, so that I test the correct state.

Reply:
@jaikiran The PR contains everything needed for the fix. It passes okay on my laptop but fails when tested against the CI.

Reply:
@mmusgrov Did you run the integration-tests/main/src/test/java/io/quarkus/it/main/TransactionITCase.java ? 
```
./mvnw -f integration-tests/main/pom.xml clean verify -Pnative
```
Does this pass okay on your laptop locally ?

Reply:
Thanks Amos. I have been running the test from the integration-tests/narayana-jta directory directly using mvn (not from the top and not using mvnw).

When I run it from the top using `./mvnw -f integration-tests/main/pom.xml clean verify -Dnative` the test fails in the same way as on CI.

So I still have the same problem as I stated in my [earlier comment](https://github.com/quarkusio/quarkus/pull/10429#discussion_r450245296). But now that I can reproduce the failure on my laptop (using your hint) I should be able to make some more progress.